crates/llms/src/chat/nsql/mod.rs

```diff
@@ -36,10 +57,28 @@ pub trait SqlGeneration: Sync + Send {
 
 /// Default system prompt for SQL code generation.
 #[must_use]
-pub fn create_prompt(query: &str) -> String {
-    format!(
+pub fn create_prompt(query: &str, ctx: &QueryGenerationContext) -> String {
+    let mut prompt = format!(
         r#"Task: Write a SQL query to answer this question: _\"{query}\"_. Instruction: Return only valid SQL code, nothing additional, don't wrap it in ```. Columns with capitals must be quoted. For tables with schemas and catalogs '"catalog"."schema"."table"' not '"catalog.schema.table"'."#
-    )
+    );
+
+    if !ctx.failed_attempts.is_empty() {
+        let failed_atttempts_str = format!("\nUse incorrectly written SQL queries and associated errors below to ensure that the new query avoids repeating the same mistakes or generating identical queries:\n\n{}", failed_attempts_formatted(&ctx.failed_attempts));
+        prompt.push_str(&failed_atttempts_str);
+    }
+
+    prompt
+}
+
+fn failed_attempts_formatted(attempts: &Vec<FailedAttempt>) -> String {
```

Comment:
We should see if table representation works better, keep this version for simplicity and performance consideration (less payload and faster conversion)